segmentstore/server/src/main/java/io/pravega/segmentstore/server/CachePolicy.java

```diff
@@ -22,11 +22,11 @@
     /**
      * Unless specified via the constructor, defines the default value for {@link #getMaxUtilization()}.
      */
-    public static final double DEFAULT_TARGET_UTILIZATION = 0.85;
+    public static final double DEFAULT_TARGET_UTILIZATION = 0.75;
```

Comment:
With the old implementation it really wasn't a big deal if we exceeded the maximum cache size. With this one we have a hard upper bound on how much data we can store in the cache - it was one of its design principles (to improve our stability).

After talking to Tom (in some of the comments on this PR) we decided it was safer to reduce these targets so that we can safely react to sudden burst in ingestion rates. For example, suppose we are sitting somewhere at the throttling limit, and we get a sudden 256MB influx of data. For a 3GB cache, 256MB is 8%, which takes us dangerously close to 100% if we kept the old values around (throttle begins at 90%, max at 98%). This would have been exacerbated if during that same time we had to perform a container recovery which would put even more pressure on the cache.

Lowering them reduces the likelihood that either one or a combination of 2 or more factors could lead to `CacheFullException`. We have tweaked this a bit in various perf tests that we've run recently and noticed a good improvement in the stability of the cluster by lowering them (stability: `CacheFullException` causes at least one Segment Container to shut down (gets evicted); the Controller will restart it in place, which may or may not fix the problem depending on how much Cache is available at that time).

Reply:
One potential, non-trivial thing we could do to avoid such discussions around the target utilization of the cache is to treat it as a real cache with say an LRU policy (or whatever other policy, I'm not stuck on LRU) and if it fills up and we have to evict data that isn't in Tier 2 yet, then we get it from BK.

Such a change implies that we would read from BK in a case that isn't only recovery, and would indeed require adding such a code path. This path would in principle not be triggered very often if the cache is dimensioned properly,  and potentially leads to a more efficient utilization of the cache space.

If you think this is worth exploring, then let's capture it in an issue.

Reply:
#4438